### ISSUE TITLE
Feature request: option to accept non-match

### ISSUE THREAD

**Issue by javrucebo** [hadley/rvest#82]
With 0.2.0.900 you have changed behaviour in case of non-match to throw error.
There are cases when you parse a document where a node might not be optional and it is perfectly fine to have an empty result. It should be possible to give an additional parameter to for example `html_node` not to throw an error/warning if there is no match.

Subsequent application of further filters should also accept the empty result, such that something like

`html_node(".id", throwError=FALSE) %>% html_text `
also works if there is no match.

**Comment by hadley** [comment]
In that case, why can't you use `html_nodes()` ?

**Comment by javrucebo** [comment]
You are indeed right. If the document has 0 or 1 occurence `html_nodes` does the job. I was too much focused on the `html_node` which worked nicely in previous `rvest` version with no-match.

The challenge is then if I want only 1 (the first) occurence (or none if no match exists).
`html_nodes` gives me all. Using `%>% extract2(1)` only works if there is minimum 1 match. 
`html_node` would give me the first occurence (but again a warning if there are several matches.

So basically the document could have any number of matches (including zero) and I would like to have: Give me zero or one (first) match. Do not issue errors or warnings otherwise.

**Comment by hadley** [comment]
Hmmm, maybe `html_node_first()`?

**Comment by gergness** [comment]
I think the original feature request would still be useful. I was trying to update the Zillow demo, but currently one of the houses doesn't have a lot size item. I can't use `html_nodes()` or `html_node_first()` because I want to keep sqft in line with the other data.

I've tried to save the page as it exists now, so hopefully this is reproducible.

``` r

url <- "https://dl.dropboxusercontent.com/u/2019891/zillow_example.html"

houses <- read_html(url) %>%
  html_nodes("article")

# Year works okay because there is exactly one per house
year_build <- houses %>%
  html_node(".built-year") %>%
  html_text() %>% 
  tidyr::extract_numeric()

# But the 11th house is missing .lot-size, so I can't get one item for each house
sqft <- houses %>%
  html_node(".lot-size") %>%
  html_text()
#> Error: No matches

sqft <- houses %>%
  html_nodes(".lot-size") %>%
  html_text()

length(sqft)
#> [1] 24

data.frame(year = year_build, sqft = sqft)
#> Error in data.frame(year = year_build, sqft = sqft): arguments imply differing number of rows: 25, 24
```

**Comment by javrucebo** [comment]
Dedicated function `html_node_first` sound good for me.

The issue which user gergness describes can be covered by iterating in parallel through the houses:

```r
getYearAndSqft <- function(node) {
  year_build <- node %>%
    html_nodes(".built-year") %>%
    html_text %>%
    tidyr::extract_numeric()
  sqft <- node %>% 
    html_nodes(".lot-size") %>%
    html_text
  list(year_build=year_build,sqft=sqft)
}

df <- houses %>% 
  lapply(getYearAndSqft) %>%
  do.call(rbind, .) %>% 
  as.data.frame

> head(df, n=13) %>% tail(n=5)
#    year_build                sqft
# 9        1992     â€¢ 0.36 ac lot
# 10       1995  â€¢ 6,534 sqft lot
# 11       2014                    
# 12       1980     â€¢ 0.40 ac lot
# 13       1967     â€¢ 0.26 ac lot
```
Also here if you would get more than one year or sqft back per house you'd have duplicated rows from using `html_nodes` - which would be needed to be taken care of separately. Having a function like `html_node_first` would make this more canonical.

**Comment by hadley** [comment]
It's a little inefficient currently, but you can use `dplyr::first()` + `purrr::map()` for problems like this:

```R
library(rvest)
library(dplyr)
library(purrr)

url <- "https://dl.dropboxusercontent.com/u/2019891/zillow_example.html"
houses <- read_html(url) %>% html_nodes("article")

houses %>%
  map_dbl(. %>% 
    html_nodes(".lot-size") %>% 
    html_text() %>% 
    readr::parse_number() %>% 
    first()
  )
```

With the right additional infrastructure (i.e. some object that represents a missing node), you'd be able to do:

```R
houses %>%
  map_chr(. %>% html_nodes(".lot-size") %>% first() %>% html_text()) %>% 
  readr::parse_number())
```

But `html_node_first()` (or maybe `html_node_nth()`?) would still be worthwhile because it would let you write:

``R
houses %>%
  html_node_nth(".lot-size", 1) %>% 
  html_text() %>% 
  readr::parse_number())
```

**Comment by gergness** [comment]
Awesome, I definitely like that last syntax better. But if there's infrastructure for empty nodes, why wouldn't `html_node()` use it?

```
houses %>%
  html_node(".lot-size") %>% 
  html_text() %>% 
  readr::parse_number())
```

Or maybe analogous to `html_attr()` there could be a `default` option?

```
# Current behavior
  html_node(".lot-size", default = stop())

# Use empty node
html_node(".lot-size", default = empty_node()) 
```

`html_node_nth()` would still be useful if you knew that there were possibly more than 1 match, as there would be no warning when taking the first, and you could also get the other ones.

**Comment by hadley** [comment]
Should be handled with new xml2 missing node tools.
